godot-core/src/engine/script_instance.rs

```diff
@@ -318,32 +259,276 @@ pub fn create_script_instance<T: ScriptInstance>(rs_instance: T) -> *mut c_void
     }
 }
 
+/// Mutable/exclusive reference guard for a `T` where `T` implements [`ScriptInstance`].
+///
+/// This can be used to access the base object of a [`ScriptInstance`], which in turn can be used to make reentrant calls to engine APIs.
+///
+/// # Example
+/// ```no_run
+/// use godot::prelude::*;
+/// use godot::builtin::meta::{MethodInfo, PropertyInfo};
+/// use godot::engine::{ScriptInstance, SiMut, ScriptLanguage, Script};
+/// use godot::sys;
+///
+/// struct ExampleScriptInstance;
+///
+/// impl ScriptInstance for ExampleScriptInstance {
+///     type Base = Object;
+///
+///     fn call(
+///         mut this: SiMut<Self>,
+///         method: StringName,
+///         args: &[&Variant],
+///     ) -> Result<Variant, sys::GDExtensionCallErrorType>{
+///         this.base_mut().emit_signal("gameover".into(), &[]);
+///
+///         Ok(Variant::nil())
+///     }
+///
+///     fn class_name(&self) -> GString { todo!() }
+///     fn set_property(_: SiMut<'_, Self>, _: godot::prelude::StringName, _: &godot::prelude::Variant) -> bool { todo!() }
+///     fn get_property(&self, _: godot::prelude::StringName) -> Option<godot::prelude::Variant> { todo!() }
+///     fn get_property_list(&self) -> Vec<PropertyInfo> { todo!() }
+///     fn get_method_list(&self) -> Vec<MethodInfo> { todo!() }
+///     fn is_placeholder(&self) -> bool { todo!() }
+///     fn has_method(&self, _: godot::prelude::StringName) -> bool { todo!() }
+///     fn get_script(&self) -> &Gd<Script> { todo!() }
+///     fn get_property_type(&self, _: godot::prelude::StringName) -> VariantType { todo!() }
+///     fn to_string(&self) -> GString { todo!() }
+///     fn get_property_state(&self) -> Vec<(godot::prelude::StringName, godot::prelude::Variant)> { todo!() }
+///     fn get_language(&self) -> Gd<ScriptLanguage> { todo!() }
+///     fn on_refcount_decremented(&self) -> bool { todo!() }
+///     fn on_refcount_incremented(&self) { todo!() }
+///     fn property_get_fallback(&self, _: godot::prelude::StringName) -> Option<godot::prelude::Variant> { todo!() }
+///     fn property_set_fallback(_: SiMut<'_, Self>, _: godot::prelude::StringName, _: &godot::prelude::Variant) -> bool { todo!() }
+/// }
+/// ````
+pub struct SiMut<'a, T: ScriptInstance> {
+    mut_ref: &'a mut T,
+    cell: Pin<&'a GdCell<T>>,
+    base_ref: &'a Base<T::Base>,
+}
+
+impl<'a, T: ScriptInstance> SiMut<'a, T> {
+    fn new(
+        cell: Pin<&'a GdCell<T>>,
+        cell_guard: &'a mut MutGuard<T>,
+        base_ref: &'a Base<T::Base>,
+    ) -> Self {
+        let mut_ref = cell_guard.deref_mut();
+
+        Self {
+            mut_ref,
+            cell,
+            base_ref,
+        }
+    }
+}
```

Comment:
Can you merge this with the lower `impl` block? You can leave `new` undocumented.

Reply:
yeah, 😅 not sure what happened there.